crates/ide-assists/src/handlers/convert_let_else_to_match.rs

```diff
@@ -0,0 +1,441 @@
+use syntax::ast::{edit::AstNodeEdit, AstNode, HasName, LetStmt, Pat};
+use syntax::T;
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+/// Gets a list of binders in a pattern, and whether they are mut.
+fn binders_in_pat(pat: &Pat) -> Option<Vec<(String, bool)>> {
+    use Pat::*;
+    match pat {
+        IdentPat(p) => {
+            let ident = p.name()?.text().to_string();
+            let ismut = p.ref_token().is_none() && p.mut_token().is_some();
+            let mut res = vec![(ident, ismut)];
+            if let Some(inner) = p.pat() {
+                res.extend(binders_in_pat(&inner)?);
+            }
+            Some(res)
+        }
```

Comment:
This could be const reference, that is an ident pat that matches `None` for example without binding a name. You can check that with `Semantics::resolve_bind_pat_to_const` (note this check is only needed if there is no `mut`, `ref` and inner pattern)

Reply:
I think it's also needed if there's only an inner pattern? like this
```rust
let a @ None = g() else { return };
```
->
```rust
let a = match g() {
    a @ None => a,
    _ => return,
};
```

Reply:
The inner pattern here is again an IdentPat so that should be already covered by the recursive nature, so the `a` is added since its not a const pat, then we recurse into the inner pattern, see the `None` which will be skipped as that is a const pattern.

Reply:
I've updated the code to check for the inner pattern.